backend/dataall/modules/dataset_sharing/services/data_sharing_service.py

```diff
@@ -226,3 +275,124 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+        try:
+            # Execute the query to get the DatasetLock object
+            dataset_lock = (
+                session.query(DatasetLock)
+                .filter(
+                    and_(
+                        DatasetLock.datasetUri == dataset_uri,
+                        ~DatasetLock.isLocked
+                    )
+                )
+                .with_for_update().one()
+            )
+
+            # Check if dataset_lock is not None before attempting to update
+            if dataset_lock:
+                # Update the attributes of the DatasetLock object
+                dataset_lock.isLocked = True
+                dataset_lock.acquiredBy = share_uri
+
+                session.commit()
+                return True
+            else:
+                log.info("DatasetLock not found for the given criteria.")
+                return False
+
+        except Exception as e:
+            session.rollback()
+            log.error("Error occurred while acquiring lock:", e)
+            return False
+
+    @staticmethod
+    def acquire_lock_with_retry(dataset_uri, session, share_uri):
+        for attempt in range(MAX_RETRIES):
+            try:
+                log.info(f"Attempting to acquire lock for dataset {dataset_uri} by share {share_uri}...")
+                lock_acquired = DataSharingService.acquire_lock(dataset_uri, session, share_uri)
+                if lock_acquired:
+                    return True
+
+                log.info(
+                    f"Lock for dataset {dataset_uri} already acquired. Retrying in {RETRY_INTERVAL} seconds...")
+                sleep(RETRY_INTERVAL)
+
+            except Exception as e:
+                log.error("Error occurred while retrying acquiring lock:", e)
+                return False
+
+        log.info(f"Max retries reached. Failed to acquire lock for dataset {dataset_uri}")
+        return False
+
+    @staticmethod
+    def release_lock(dataset_uri, session, share_uri):
+        """
+        Releases the lock on the dataset identified by dataset_uri.
+
+        Args:
+            dataset_uri: The ID of the dataset for which the lock is being released.
+            session (sqlalchemy.orm.Session): The SQLAlchemy session object used for interacting with the database.
+            share_uri: The ID of the share that is attempting to release the lock.
+
+        Returns:
+            bool: True if the lock is successfully released, False otherwise.
+        """
+        try:
+            log.info(f"Releasing lock for dataset: {dataset_uri} last acquired by share: {share_uri}")
+            query = (
+                session.query(DatasetLock)
+                .filter(
+                    and_(
+                        DatasetLock.datasetUri == dataset_uri,
+                        DatasetLock.isLocked == True,
+                        DatasetLock.acquiredBy == share_uri
+                    )
+                )
+                .with_for_update().one()
+            )
```

Comment:
@anushka-singh - sorry for all the back and forth ...

with the latest changes to add `.with_for_update().one()` I am now receiving error: `"Arguments: (AttributeError("'DatasetLock' object has no attribute 'update'"),)"`


I think for we have 2 options:

- (1) For `def release_lock()` we can handle the update the same as  for acquire lock 
```
            dataset_lock = (
                session.query(DatasetLock)
                .filter(
                    and_(
                        DatasetLock.datasetUri == dataset_uri,
                        DatasetLock.isLocked == True,
                        DatasetLock.acquiredBy == share_uri
                    )
                )
                .with_for_update().one()
            )
           if dataset_lock:
                dataset_lock.isLocked = False
                dataset_lock.acquiredBy = ''
...
```

- (2) OR perform query and update all in 1:
```
            query = (
                session.query(DatasetLock)
                .filter(
                    and_(
                        DatasetLock.datasetUri == dataset_uri,
                        DatasetLock.isLocked == True,
                        DatasetLock.acquiredBy == share_uri
                    )
                )
                .update(
                {
                    "isLocked": False,
                    "acquiredBy": ''
                },
                )
            )
``` (and likely can handle `acquire_lock()` similarly)

Reply:
Hmm, that is really weird. I did not come across this issue in my testing!

But I have pushed out another revision with solution (1). Please LMK if its fixed now. I will also try it out again on my end.

Reply:
will do - trying again now